Socket.io client instrumentation (#304)
signalfx/splunk-otel-js-web: docs/Instrumentations.md

|---|---|---|
|`http.url`|`string`|Websocket URL.|
|`response_content_length`|`number`|Payload size in bytes.|

> ## Instrumentation: Socket.io client

Reviewer: Maybe mention that we support all versions 1-4.

> This instrumentation generates spans from messages sent using [socket.io](https://socket.io/) client library. The generated spans conform to [OpenTelemetry specifications on messaging systems](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/messaging.md). This instrumentation is disabled by default as it requires a bit more advanced setup.

Reviewer 2: ```suggestion
This instrumentation generates spans from messages sent using the [socket.io](https://socket.io/) client library. Generated spans conform to the [OpenTelemetry specifications on messaging systems](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/messaging.md). This instrumentation is disabled by default.
```

> > Enabling websocket instrumentation when using socket.io isn't generally recommended.

Reviewer 2: ```suggestion
> Enabling websocket instrumentation when using socket.io isn't generally recommended.
```

Can we be more specific about why it shouldn't be done?

Author: It's mostly due to the reason this instrumentation was made - websocket is basically just a tcp connection so instrumenting it just shows some data went in, some data came out, so it's kinda useless when a better instrumentation that includes what type of message was sent also exists. But technically there's nothing conflicting if someone does enable both

Reviewer 2: When would one prefer websocket over socket.io? It's just because of the messaging data model? If that's the case, why doesn't the websocket instrumentation follow it? Trying to understand if we can replace this note with a more useful recommendation based on instrumentation needs.

Author: Hmm so issue that websocket instrumentation has is that it can't really give a good [span name as wanted by the messaging spec](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/messaging.md#span-name) - since the data can be in any structure as defined by the user (can be a random string `hello`, can be a string with structure that only the dev knows eg `hello/user=1`, can be stringified json, can be raw bytes or [msgpack](https://msgpack.org/), ...) there isn't a good way to extract a name for what data is being sent

Meanwhile socket.io has a low cardinality event name that both client and server use to determine what to do with the message, so we can generate spans with `event_name send/receive` name.

Socket.io uses websocket as once of the transport methods internally so it would cause spans from both instrumentations, where the inner (websocket data sent) span is less useful. Tho socket.io can also do http-polling or comet for sending data when websocket is unavailable, so http spans would be generated then, so I can see removing this warning too since the lower level communication method can appear anyway

Reviewer 2: Then

```suggestion
```

### Setup

#### Socket.io client library served by Socket.io server

> When using [the standalone build server by socket.io server](https://socket.io/docs/v4/client-installation/#standalone-build), instrumentation can be just enabled by passing `true` to configuration.

Reviewer 2: ```suggestion
When using [the standalone build by socket.io server](https://socket.io/docs/v4/client-installation/#standalone-build), you can enable instrumentation by passing `true` to the configuration setting, like in the following snippet:
```

```html
<script src="/location/to/splunk-otel-web.js"></script>
<script>
  SplunkRum.init({
    // ...
    instrumentations: {
      socketio: true
    }
  });
</script>
<script src="/socket.io/socket.io.js"></script>
```

#### Bundled socket.io client with NPM Splunk RUM installation

> When using both `@splunk/otel-web` and `socket.io-client` by bundling together, socket.io client should be passed to the instrumentation via `target` config option:

Reviewer 2: ```suggestion
When using both the `@splunk/otel-web` and the `socket.io-client` packages bundled together, pass the socket.io client to the instrumentation using the `target` config option:
```

```js
import SplunkOtelWeb from '@splunk/otel-web';
import { io } from 'socket.io-client';

SplunkRum.init({
  // ...
  instrumentations: {
    socketio: {
      target: io,
    },
  },
});
```

#### Bundled socket.io client with CDN Splunk RUM installation

> When using CDN distribution of Splunk RUM, socket.io instrumentation should be enabled expecting a global object and inside the bundle `io` function should be exposed as `window.io` 

Reviewer 2: ```suggestion
When using the CDN distribution of Splunk RUM, enable the socket.io instrumentation using a global object and exposing the `io` function as `window.io`, as in the following example:
```

Is the rewording correct?

Author: hmm "enable the socket.io instrumentation using a global object" kinda means nothing, maybe

```suggestion
When using the CDN distribution of Splunk RUM, enable the socket.io instrumentation and expose the `io` function as `window.io`, as in the following example:
```

Reviewer 2: +1

```html
<script src="/location/to/splunk-otel-web.js"></script>
<script>
  SplunkRum.init({
    // ...
    instrumentations: {
      socketio: true
    }
  });
</script>
<script src="/app.min.js"></script>
```

> 

Reviewer 2: The JS snippet that follows is the content of what? `app.min.js`?

Author: Yes, or at least the source files that eventually get bundled together into the minified `app.min.js`. Maybe best label for it is "where you use socket.io"

```js
import { io } from 'socket.io-client';

window.io = io;

const socket = io();
// ...
```

A different global variable name can be used by specifying it as target:

```js
SplunkRum.init({
  // ...
  instrumentations: {
    socketio: {
      target: 'socketIoClient',
    },
  },
});

> // Then expose the io object in your bundle by doing:

Reviewer 2: ```suggestion
// Expose the io object in your bundle
```

window.socketIoClient = io;
```

### Spans

> Messages sent between socket.io client and server are sent as `EVENT_NAME send` (client -> server) or `EVENT_NAME receive` (server -> client) spans with the following tags:

Reviewer 2: ```suggestion
Messages sent between socket.io clients and servers are `EVENT_NAME send` (client to server) or `EVENT_NAME receive` (server to client) spans with the following tags:
```
